static/app/components/events/interfaces/exception.tsx

```diff
@@ -1,97 +1,105 @@
-import {Component} from 'react';
+import {useState} from 'react';
 
 import EventDataSection from 'app/components/events/eventDataSection';
 import CrashContent from 'app/components/events/interfaces/crashContent';
 import CrashActions from 'app/components/events/interfaces/crashHeader/crashActions';
 import CrashTitle from 'app/components/events/interfaces/crashHeader/crashTitle';
 import {isStacktraceNewestFirst} from 'app/components/events/interfaces/stacktrace';
 import {t} from 'app/locale';
-import {ExceptionType} from 'app/types';
+import {ExceptionType, Group} from 'app/types';
 import {Event} from 'app/types/event';
 import {STACK_TYPE, STACK_VIEW} from 'app/types/stacktrace';
 
-const defaultProps = {
-  hideGuide: false,
-};
-
 type Props = {
   event: Event;
   type: string;
   data: ExceptionType;
   projectId: string;
-} & typeof defaultProps;
-
-type State = {
-  stackView: STACK_VIEW;
-  stackType: STACK_TYPE;
-  newestFirst: boolean;
+  hasGroupingTreeUI: boolean;
+  groupingCurrentLevel?: Group['metadata']['current_level'];
+  hideGuide?: boolean;
 };
 
-class Exception extends Component<Props, State> {
-  static defaultProps = {
-    hideGuide: false,
-  };
+function Exception({
+  event,
+  type,
+  data,
+  projectId,
+  groupingCurrentLevel,
+  hasGroupingTreeUI,
+  hideGuide = false,
+}: Props) {
+  const [stackView, setStackView] = useState<STACK_VIEW>(
+    data.hasSystemFrames ? STACK_VIEW.APP : STACK_VIEW.FULL
+  );
+  const [stackType, setStackType] = useState<STACK_TYPE>(STACK_TYPE.ORIGINAL);
+  const [newestFirst, setNewestFirst] = useState(isStacktraceNewestFirst());
 
-  state: State = {
-    stackView: this.props.data.hasSystemFrames ? STACK_VIEW.APP : STACK_VIEW.FULL,
-    newestFirst: isStacktraceNewestFirst(),
-    stackType: STACK_TYPE.ORIGINAL,
-  };
+  const eventHasThreads = !!event.entries.find(entry => entry.type === 'threads');
 
-  handleChange = (newState: Partial<State>) => {
-    this.setState(prevState => ({
-      ...prevState,
-      ...newState,
-    }));
-  };
+  // in case there are threads in the event data, we don't render the
+  // exception block.  Instead the exception is contained within the
+  // thread interface.
+  if (eventHasThreads) {
+    return null;
+  }
 
-  render() {
-    const eventHasThreads = !!this.props.event.entries.find(
-      entry => entry.type === 'threads'
-    );
+  function handleChange({
```

Comment:
noob question, won't this function be created on every render? https://stackoverflow.com/a/56459512/1544347

it appears you have a lot of nested helper functions like that, I wonder if that affects performance for large stacktraces

in either case the refactor to class-based components makes this diff hard to read

Reply:
yes, it will. "does it impact performance in a significant way?" it depends on the case, but I would say that in most of the cases no. We can use some performance optimization functions if we notice that the performance of large stack traces is bad